### PR TITLE
create contiv policy based on kubernetes default isolation policy.

### PR DESCRIPTION
**Pull request by rchirakk**
Initial part of network policy implementation
move some of the util functions to k8sutil

Signed-off-by: Ranjith <rchirakk@users.noreply.github.com>

### CODE REVIEW

**Comment by rchirakk** [comment]
@gkvijay @DivyaVavili PTAL

File: netmaster/k8snetwork/networkpolicy.go [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+		}
+		// ignore other events
+	}
+}
+
+// InitK8SServiceWatch monitor k8s services
+func InitK8SServiceWatch(isLeader func() bool) error {
+	npLog = log.WithField("k8s", "netpolicy")
+
+	contivClient, err := client.NewContivClient("http://localhost:9999")
```
Comment (gkvijay) [inline netmaster/k8snetwork/networkpolicy.go]: The listen port is configurable. Should pick it up from the listenURL

File: netmaster/k8snetwork/networkpolicy.go [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+	npLog.Infof("process [%s] network policy  %+v", opCode, np)
+
+	switch opCode {
+	case watch.Added, watch.Modified:
+	case watch.Deleted:
+	}
+}
+
+func (k8sNet *k8sContext) processK8sEvent(opCode watch.EventType, eventObj interface{}) {
+	if k8sNet.isLeader() != true {
```
Comment (DivyaVavili) [inline netmaster/k8snetwork/networkpolicy.go]: For my understanding, why are the isLeader checks made in many places?

File: netmaster/k8snetwork/networkpolicy.go [hunk 3]
```diff
@@ -0,0 +1,424 @@
+package networkpolicy
+
+import (
+	"encoding/json"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"github.com/contiv/contivmodel/client"
+	"github.com/contiv/netplugin/utils/k8sutils"
+	"github.com/rchirakk/client-go/kubernetes"
```
Comment (DivyaVavili) [inline netmaster/k8snetwork/networkpolicy.go]: @rchirakk: Looks like the networkpolicy related changes is checked into the official client-go. Please check and move to official client-go.

File: netmaster/k8snetwork/networkpolicy.go [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+	if _, err := k8sNet.contivClient.NetworkGet(defaultTenantName, nwName); err == nil {
+		return nil
+	}
+
+	if err := k8sNet.contivClient.NetworkPost(&client.Network{
+		TenantName:  defaultTenantName,
+		NetworkName: nwName,
+		Subnet:      defaultSubnet,
+		Encap:       "vxlan",
```
Comment (rchirakk) [inline netmaster/k8snetwork/networkpolicy.go]: vxlan is by default like in netctl net create.
user need to create vlan network prior to applying network policy

File: netmaster/k8snetwork/networkpolicy.go (hunk 2 above)
Comment (rchirakk) [inline netmaster/k8snetwork/networkpolicy.go]: to avoid epg creation from multiple masters if exists.

File: netmaster/daemon/daemon.go [hunk 5]
```diff
@@ -326,6 +327,19 @@ func (d *MasterDaemon) becomeFollower() {
...
+// InitServices init watch services
+func (d *MasterDaemon) InitServices() {
+	if d.ClusterMode == "kubernetes" {
+		isLeader := func() bool {
+			if d.currState == "leader" {
+				return true
+			}
+			return false
+		}
+		networkpolicy.InitK8SServiceWatch(d.ListenURL, isLeader)
```
Comment (rchirakk) [inline netmaster/daemon/daemon.go]: it is to POST config, so listenURL is used.

File: netmaster/daemon/daemon.go [hunk 6]
```diff
@@ -326,6 +327,19 @@ func (d *MasterDaemon) becomeFollower() {
 	go d.runFollower()
 }
 
+// InitServices init watch services
+func (d *MasterDaemon) InitServices() {
+	if d.ClusterMode == "kubernetes" {
+		isLeader := func() bool {
+			if d.currState == "leader" {
```
Comment (dseevr) [inline netmaster/daemon/daemon.go]: This can be simplified to `return d.currState == "leader"`

File: netmaster/k8snetwork/networkpolicy_test.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+		resp, err := handlerFunc(w, r, mux.Vars(r))
+		if err != nil {
+			// nptLog error
+
+			// Send HTTP response
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		} else {
+			// Send HTTP response as Json
+			err = writeJSON(w, http.StatusOK, resp)
+			if err != nil {
```
Comment (dseevr) [inline netmaster/k8snetwork/networkpolicy_test.go]: Missing body for this if statement?

File: netmaster/k8snetwork/networkpolicy_test.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+type httpAPIFunc func(w http.ResponseWriter, r *http.Request, vars map[string]string) (interface{}, error)
+
+// Simple Wrapper for http handlers
+func makeHTTPHandler(handlerFunc httpAPIFunc) http.HandlerFunc {
+	// Create a closure and return an anonymous function
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Call the handler
+		resp, err := handlerFunc(w, r, mux.Vars(r))
+		if err != nil {
+			// nptLog error
```
Comment (dseevr) [inline netmaster/k8snetwork/networkpolicy_test.go]: Missing the `ntpLog.Fatalf` here

File: netmaster/daemon/daemon.go (hunk 6 above)
Comment (rchirakk) [inline netmaster/daemon/daemon.go]: 👍

File: netmaster/k8snetwork/networkpolicy_test.go (hunk 8 above)
Comment (rchirakk) [inline netmaster/k8snetwork/networkpolicy_test.go]: removed unused code

File: netmaster/k8snetwork/networkpolicy_test.go (hunk 7 above)
Comment (rchirakk) [inline netmaster/k8snetwork/networkpolicy_test.go]: removed unused code

**Comment by rchirakk** [comment]
updated godep.
@dseevr @DivyaVavili  PTAL

**Comment by DivyaVavili** [comment]
LGTM

**Pull request merged**